### PR TITLE
Update Intents "Distribution AppStore" configuration to match other extensions

### PR DESCRIPTION
**Pull request by mokagio**
### Fix
While checking what discussed [here](https://github.com/Automattic/simplenote-ios/pull/1375#discussion_r693999625), I noticed that the entitlements for the "Distribution AppStore" build configuration was pointing to the Alpha build. That was inconsistent with the other apps, but most importantly incorrect: "Distribution AppStore" is the build configuration we use to App Store builds and needs to point to a valid app group (the Alpha one exists only in the Enterprise account).

### Test
In Xcode, verify the "Signing & Capabilities > App Groups" are consistent between extensions

**Share**

Screen Shot 2021-08-25 at 3 10 49 pm

**Widgets**

Screen Shot 2021-08-25 at 3 18 23 pm

**Intents**

Screen Shot 2021-08-25 at 3 11 06 pm

### Review
Only one developer required to review these changes, but anyone can perform the review.

### Release
These changes do not require release notes.

### CODE REVIEW

**Review by mokagio — COMMENTED** [review]
While reviewing these changes, I got scared for a minute or two because, from Xcode, it looks like the app group for the "Distribution AppStore" is the development one 😱 

image

Looking at the `.entitlements` file, though, I'm pretty sure that's not the case

https://github.com/Automattic/simplenote-ios/blob/8ef355d4133ce0db9acde631d91e466685ab5218/Simplenote/Simplenote.entitlements#L9-L12

I'm not sure how to prove it, though, other than indirectly by ensuring that the share extension works in production, which it does.

I downloaded the latest archive and looked for the entitlements file in there with

```
find /path/to/xcarchive -name "*.entitlements"
```

But the only result I find is the one for the share extension 🤔 @Automattic/owl-team any ideas on how to be absolutely sure about this?

**Review by charliescheer — APPROVED** [review]
Looks good to me

**Comment by mokagio** [comment]
Thanks @charliescheer. 

I'd like to wait an hear for input from someone in @Automattic/owl-team before merging.

If this suddenly becomes a blocker for your work, though, feel free to take over and merge. We'll fix as we go if needed.

**Comment by AliSoftware** [comment]
> While reviewing these changes, I got scared for a minute or two because, from Xcode, it looks like the app group for the "Distribution AppStore" is the development one 😱
> 
> image

I'm pretty sure this glitch in Xcode UI is coming from the fact that we use the same `.entitlements` **file** for those build configs, and Xcode isn't smart enough when showing the "Signing & Capabilities" tab to realize that even if they're the same file, the variable `$(CFBundleIdentifier)` in them with **expand to something different** and thus lead to different entitlements in practice.

So Xcode sees this in the target's Build Settings:
image
and thus assumes it's gonna be the same entitlements for Debug, Release and Distribution App Store, decodes/interprets the `.entitlements` file only once (in the context of 1 of the 3 build configs arbitrarily, apparently Debug) and builds that UI showing the App Groups using that.

> I downloaded the latest archive and looked for the entitlements file in there with
> 
> ```
> find /path/to/xcarchive -name "*.entitlements"
> ```
> 
> But the only result I find is the one for the share extension 🤔 

This actually highlights a separate configuration problem: bare `*.entitlements` are not supposed to be embed in the final bundle, and indeed if you look at the target membership for all the `*.entitlements` files, none of them have any target checked… except the `SimplenoteShare-AppStore.entitlements` one. That's confirmed by this warning we get during archive:

image

Instead, we should remove that `*.entitlements` file from the target, because entitlements are embeded in the final provisioning profile (`*.mobileprovision`) when you archive and bundle the app, and that's what you need to inspect.

---

I looked at the local provisioning profiles I have installed on my Mac and especially the "Simplenote AppStore*" ones (= one for app + one for share ext) and they both seem to have the correct app groups set:

```
$ pprof --name 'Simplenote AppStore' -0 | xargs -0 -n1 security cms -D -i | grep -A 3 'com.apple.security.application-groups'
				<key>com.apple.security.application-groups</key>
		<array>
				<string>group.com.codality.NotationalFlow</string>
		</array>
--
				<key>com.apple.security.application-groups</key>
		<array>
				<string>group.com.codality.NotationalFlow</string>
		</array>
```

I've started an archive build and will run the same `security cms` command on the `embed.mobileprovision` file in the final `.app` of the produced `xcarchive` to double-check, just waiting for the build to finish and I'll post an update here.

**Comment by AliSoftware** [comment]
Useful additional Apple resource about tips for checking entitlements: https://developer.apple.com/documentation/bundleresources/entitlements/diagnosing_issues_with_entitlements

**Comment by AliSoftware** [comment]
@mokagio While testing the archive on my end I noticed another warning about the Intents profile:

image

I can confirm the presence of these via this command ([here to install pprof](https://github.com/AliSoftware/pprof) if you don't have it yet):
```
$ pprof --name '/match AppStore .*NotationalFlow.Intents'/ -0 | xargs -0 pprof -i
- name: match AppStore com.codality.NotationalFlow.Intents
- uuid: 1d083f91-ce79-447a-82af-368b4181f906
- app_id_name: Simplenote Intents
- app_id_prefix: ["PZYM8XX95Q"]
- creation_date: 2021-07-05T05:17:04+00:00
- expiration_date: 2022-02-23T04:32:18+00:00
- ttl: 232
- team_ids: ["PZYM8XX95Q"]
- team_name: Automattic, Inc.
- Entitlements:
   - beta-reports-active: true
   - keychain-access-groups: ["PZYM8XX95Q.*", "4ESDVWK654.com.codality.NotationalFlow"]
   - previous-application-identifiers: ["4ESDVWK654.com.codality.NotationalFlow", "4ESDVWK654.com.codality.NotationalFlow.Share"]
   - com.apple.security.application-groups: ["group.com.codality.NotationalFlow"]
   - application-identifier: PZYM8XX95Q.com.codality.NotationalFlow.Intents
   - get-task-allow: false
   - com.apple.developer.team-identifier: PZYM8XX95Q
- 1 Developer Certificates
- 0 Provisioned Devices
- Provision all devices: false
```

The `previous-application-identifiers` key seems to be related to the fact that we transfered the app from one AppStore account to another a while ago. That key is also present in `Simplenote.entitlements` and `SimplenoteShare-AppStore.entitlements` which means we probably want to add this to our `SimplenoteIntents.entitlements` as well 🤔

**Comment by AliSoftware** [comment]
@mokagio I took the liberty to push a couple of commits to fix a couple of issues:
 - Removed the Share extensions' `entitlements` file from target membership (825a440)
 - Add the `previous-application-identifiers` entitlement that was apparently expected by the profile used for Intents but missing in entitlements file (a511619)
 - Did a big cleanup and standardization pass on the `*.entitlements` and related supporting files, so that each target use a similar convention
    - Move the `*.entitlements` and `Info.plist` files of each target in respective dedicated `Supporting Files` folder on disk to nicely isolate them
    - Renamed some `*.entitlements` files to standardize their name to `<Target>-<Config>.entitlements` for clarity and consistency
    - Updated the `INFOPLIST_FILE` and `CODE_SIGN_ENTITLEMENTS` build settings of each target (app, widget, share, intents) accordingly (670bcf0, 50b7282, 46e1594, 207f4e1)
 - Removed the unused `InfoPlist.strings` file (2f9c8ec)

I'd love for you to review those changes and commits before merging

---

After all those changes, I did a quick pass for checking:

 - ✅ Built the app in Debug config, and ensured that it didn't fail nor generate warnings on Info.plist, PCH, or entitlements
 - ✅  Did the same with "Distribution Internal", "Distribution Alpha" and "Distribution AppStore" configs
 - ❌  Couldn't test the same for the "Release" configuration because I apparently didn't have the profiles for it (`No profile for team 'Automattic, Inc. (Company)' matching 'Simplenote Release' found: Xcode couldn't find any provisioning profiles matching 'PZYM8XX95Q/Simplenote Release'.`) but I think that might be kinda expected? (and I didn't make any change to the provisioning profiles settings, only to entitlements ones, so I figure if there's really an issue here that'd be unrelated anyway)
 - ✅  Archiving then checking the `embed.mobileprovision` for included entitlements confirms the right AppGroup is used for `Distribution AppStore`, so I can confirm the setup works as expected:

<details><summary>Output of <tt>security cms -D -i …/embed.mobileprovision</tt></summary>

```
$ security cms -D -i /Users/olivier/Library/Developer/Xcode/Archives/2021-08-26/Simplenote\ 26-08-2021\ 14.19.xcarchive/Products/Applications/Simplenote.app/embedded.mobileprovision | plutil -extract Entitlements xml1 -o - -                                      
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>application-identifier</key>
	<string>PZYM8XX95Q.com.codality.NotationalFlow</string>
	<key>beta-reports-active</key>
	<true/>
	<key>com.apple.developer.applesignin</key>
	<array>
		<string>Default</string>
	</array>
	<key>com.apple.developer.associated-domains</key>
	<string>*</string>
	<key>com.apple.developer.team-identifier</key>
	<string>PZYM8XX95Q</string>
	<key>com.apple.security.application-groups</key>
	<array>
		<string>group.com.codality.NotationalFlow</string>
	</array>
	<key>get-task-allow</key>
	<false/>
	<key>keychain-access-groups</key>
	<array>
		<string>PZYM8XX95Q.*</string>
		<string>4ESDVWK654.com.codality.NotationalFlow</string>
	</array>
	<key>previous-application-identifiers</key>
	<array>
		<string>4ESDVWK654.com.codality.NotationalFlow</string>
		<string>4ESDVWK654.com.codality.NotationalFlow.Share</string>
	</array>
</dict>
</plist>
```
</details>

Note: for each first time I ran those for each config, Xcode notified me that I didn't yet have the corresponding provisioning profile locally and suggested me to download it on my Mac (which worked as expected) – proving in a way that it was using the profile for those builds as expected.

**Comment by AliSoftware** [comment]
Dammit, seems we have a `pbxproj` conflict with all my changes now 🙁  (… should have seen that comming 😢 )

I've spent way more time than I originally planned on this for today… so I think I'm gonna avoid jumping into the rabbit hole of `pbxproj` conflict resolution for now, and give it a rest.

@charliescheer @mokagio don't hestiate to give it a stab; especially I'm not sure what are the changes made or pending on the `feature/note-widget-pre-merge` target branch, and thus how best to resolve these conflicts in the first place. Maybe that's not that bad and easy enough to solve after all, but you likely have more context than I do about the other branch to help resolving and giving it the finishing touch.

**Comment by charliescheer** [comment]
Hey @AliSoftware and @mokagio  Thank you so much for helping out with this!  This PR got a little lost in the shuffle, sorry for  the delay in getting back to it.  I took care of the merge conflicts, and I can still build on a device, however I am seeing an entitlements warning when I build to device.



I'm not sure if that is something I caused while fixing the conflicts, but apart from that it seems to be good to go.  Any suggestions on what I should check out to deal with this warning?

**Comment by AliSoftware** [comment]
Thanks @charliescheer for taking care of the conflict!

> I am seeing an entitlements warning when I build to device. [...]
> I'm not sure if that is something I caused while fixing the conflicts, but apart from that it seems to be good to go. Any suggestions on what I should check out to deal with this warning?

I've fixed that same warning for `SimplenoteIntents` in a511619ab5dc695924949f3420ef1cf841d2ed6f, but didn't for the Widget one, and I think you should fix it the same way for the WidgetsExtension.

---

After fixing it in a511619ab5dc695924949f3420ef1cf841d2ed6f I remember rebuilding the project and I indeed found it surprising that I didn't get a similar warning for `SimplenoteWidgetsExtension` then – given than when I checked all the various `.entitlements` files during that cleanup, I remembered seeing that `previous-application-identifier` was present in all of them… except the WidgetExtension one indeed.

Back then I though something like "eh, maybe it's because the widget and thus its provisioning profile got created _after_ the app got transferred to our a8c account, and so didn't need have `previous-app-identifiers` in the profile, hence it not being in the entitlements for this target?"… but now I realize that if that were really the case that would have been true for the Intents target as well. I should also have [checked with `pprof`](https://github.com/Automattic/simplenote-ios/pull/1408#issuecomment-906316711) to realize that the Widget's `.mobileprovision` indeed has the entitlement – which confirms it should be added to the `.entitlement` file too.

The reason I didn't have that same warning that you had for the WidgetsExtension target is probably because "eh, Xcode." 😅 
 (I also might not have done a full _clean_ build, and as we know sometimes Xcode doesn't always update the warnings list properly in incremental builds…)

**Comment by charliescheer** [comment]
Awesome!  Thanks @AliSoftware!!  I have made that change and now I have no more warnings.  I would like to merge this tomorrow if possible.  Do you think this is good to go?

**Comment by AliSoftware** [comment]
I think there are still entitlements missing in the WidgetExtension for some configurations.

I wrote this small Ruby script using `xcodeproj` and `plist` gems and `pprof`+`security` CLI to check the various configs:

<details><summary>check-ent.rb Script</summary>

```ruby
#!/usr/bin/env ruby

require 'xcodeproj'
require 'plist'

PAI_KEY = 'previous-application-identifiers'

@profiles_memo = {}
def entitlements_for_profile(prof_name)
  @profiles_memo[prof_name] ||= begin
    # Find the Provisioning Profile path from the name, using pprof
    prof_path = `pprof -l --name '/^#{prof_name}$/'`.split("\n").first
    return nil if prof_path.nil?
    # Decode the mobileprovision file into XML Plist
    raw_xml = `security cms -D -i '#{prof_path}'`
    # Parse the XML Plist into a ruby structure
    xml = Plist.parse_xml(raw_xml)
    xml['Entitlements']
  end
end

x = Xcodeproj::Project.open('Simplenote.xcodeproj')

x.targets.each do |t|
  next if t.build_settings('Debug')['CODE_SIGN_ENTITLEMENTS'].nil?
  puts "== #{t.name} =="
  t.build_configurations.each do |c|
    # Entitlements from *.entitlements file
    entitlements = Plist.parse_xml(c.build_settings['CODE_SIGN_ENTITLEMENTS'])
    
    # Entitlements from provisioning file
    prof_name = c.build_settings['PROVISIONING_PROFILE_SPECIFIER']
    prof_ent = entitlements_for_profile(prof_name)
    
    if prof_ent.nil?
      puts " - #{c.name}: Profile not found on disk"
    elsif ((entitlements[PAI_KEY] || []) - (prof_ent[PAI_KEY] || [])).empty?
      puts " - #{c.name}: OK"
    else
      puts " - #{c.name}: KO -- #{entitlements[PAI_KEY].inspect} ⊄ #{prof_ent[PAI_KEY].inspect}"
    end
  end
end
```
</details>

And here's what it printed:

```
== Simplenote ==
 - Debug: OK
 - Release: Profile not found on disk
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
== SimplenoteShare ==
 - Debug: OK
 - Release: Profile not found on disk
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
== SimplenoteWidgetsExtension ==
 - Debug: OK
 - Release: KO -- ["4ESDVWK654.com.codality.NotationalFlow"] ⊄ nil
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: KO -- ["4ESDVWK654.com.codality.NotationalFlow"] ⊄ nil
== SimplenoteIntents ==
 - Debug: OK
 - Release: OK
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
```

Indeed when looking at the `*.entitlements` file used for "Release" and "Distribution AppStore", aka `SimplenoteWidgetsExtension-Release.entitlements` (for both), the `*.entitlements` file contains this `previous-application-identifiers` key… but when looking at the `match AppStore com.codality.NotationalFlow.Widgets` provisioning profile, this entitlement doesn't seem to be present:

```
$ pprof 89859073-b915-481d-a6f7-06ff1a0ebad7
- name: match AppStore com.codality.NotationalFlow.Widgets
- uuid: 89859073-b915-481d-a6f7-06ff1a0ebad7
- app_id_name: Simplenote Widgets
- app_id_prefix: ["PZYM8XX95Q"]
- creation_date: 2021-06-02T20:59:09+00:00
- expiration_date: 2022-02-23T04:32:18+00:00
- ttl: 265
- team_ids: ["PZYM8XX95Q"]
- team_name: Automattic, Inc.
- Entitlements:
   - beta-reports-active: true
   - com.apple.security.application-groups: ["group.com.codality.NotationalFlow"]
   - application-identifier: PZYM8XX95Q.com.codality.NotationalFlow.Widgets
   - keychain-access-groups: ["PZYM8XX95Q.*", "com.apple.token"]
   - get-task-allow: false
   - com.apple.developer.team-identifier: PZYM8XX95Q
- 1 Developer Certificates
- 0 Provisioned Devices
- Provision all devices: false
```

Not sure why the discrepency, but that means that if we try to build one of those 2 configuration we will get an error.
I couldn't try to fully build those because I don't seem to have all the provisioning profiles for the Release configuration, and in fact I wonder if the Release configuration isn't even more broken given what I see in the errors I get. But one thing for sure is that amongst the errors we can indeed see that Xcode is unhappy with that entitlement mismatch (it's even an error there):

image

I'm unsure what's the best approach here:
 - Merge this PR, and open a subsequent one to try to fix the Release configuration (as it appears the build failure isn't just for Intents but other targets have issues with this config too)
 - Consider that the Release config was always broken?
 - Try to fix this before merging that PR – I haven't checked if the profiles and certificates that I am missing on my Mac to build that config are just me not having downloaded everything, or if it's a shared error for all of us though.

**Comment by AliSoftware** [comment]
According to @mokagio in Slack DM, the `Release` configuration hasn't been used in ages and might have been broken for a long time. So I think we can merge this after all as the `Release` configuration not working is unrelated to what this PR addresses.

**Comment by charliescheer** [comment]
Awesome!  Thanks @AliSoftware  I will go ahead and merge this and get the feature moved into develop 😄

**Comment by AliSoftware** [comment]
> Awesome!  Thanks @AliSoftware  I will go ahead and merge this and get the feature moved into develop 😄 

@charliescheer We can ignore the "Release" config failure… but I think we still need to make sure the "Distribution AppStore" config works and is fixed for the "SimplenoteWidgetsExtension" target first though!!

> Indeed when looking at the *.entitlements file used for "Release" and "Distribution AppStore", aka SimplenoteWidgetsExtension-Release.entitlements (for both), the *.entitlements file contains this previous-application-identifiers key… but when looking at the match AppStore com.codality.NotationalFlow.Widgets provisioning profile, this entitlement doesn't seem to be present

**Comment by charliescheer** [comment]
>but I think we still need to make sure the "Distribution AppStore" config works and is fixed for the "SimplenoteWidgetsExtension" target first though!!

ah right, I misunderstood.  Good thing I hadn't gotten to merging yet :)

**Comment by mokagio** [comment]
_Just a note to say that I started looking into this... I'll keep you posted._

**Comment by mokagio** [comment]
@AliSoftware @charliescheer I got it to work with the changes I pushed yesterday. The Xcode UI doesn't show any more warnings in the "Signing & Capabilities" tab and Olivier's script returns all `OK` 🙂

```
== Simplenote ==
 - Debug: OK
 - Release: OK
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
== SimplenoteShare ==
 - Debug: OK
 - Release: OK
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
== SimplenoteWidgetsExtension ==
 - Debug: OK
 - Release: OK
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
== SimplenoteIntents ==
 - Debug: OK
 - Release: OK
 - Distribution Internal: OK
 - Distribution Alpha: OK
 - Distribution AppStore: OK
 ```

Key in getting this to work was removing `previous-application-identifiers` entry from the widget extension entitlements. I added my rationale in the commit message: 660d825b00b2ed9d2461b274642d4782613c2403.

In the commit message, I also mention that I'm surprised that the provisioning profile and entitlements for the, newest, Intents target have the the entry. Only afterwards I noticed that @AliSoftware added them to the file, here 825a440243857a1ef140883de96b7d3660cb87e3. 

I tried removing them from both Intents and Widgets. The Xcode UI is still clear, the script returns `OK`, and I can successfully archive and validate with the store.

My theory is that we don't need that value in the new targets because they are new and never had a chance to be installed on a device with a different application identifier. I and found the PR that added those values. I left [a comment](https://github.com/Automattic/simplenote-ios/pull/180#issuecomment-916637316) there, too, asking for extra context.

At this point, I think we should ship without those entries, on the basis that the tests we put in place pass and that, usually, less configuration and overhead is better – code is liability.

File: SimplenoteWidgets/Supporting Files/SimplenoteWidgetsExtension-Debug.entitlements
Comment (AliSoftware) [inline SimplenoteWidgets/Supporting Files/SimplenoteWidgetsExtension-Debug.entitlements]: @mokagio Given your explanation from last comment (which I tend to agree with), shouldn't we remove the key in this `SimplenoteWidgetsExtension-Debug.entitlements` file too, to align with the fact that it's not here in other `SimplenoteWidgetsExtension-<Config>.entitlements` file?

```suggestion
```

**Review by AliSoftware — APPROVED** [review]
I agree with @mokagio 's rationale and the reasoning 👍 

In fact [I had the same reasoning about "the previous-app-id not being needed because the target was created way after app transfer"](https://github.com/Automattic/simplenote-ios/pull/1408#issuecomment-910166188) back then too:

> Back then I though something like "eh, maybe it's because the widget and thus its provisioning profile got created after the app got transferred to our a8c account, and so didn't need have previous-app-identifiers in the profile, hence it not being in the entitlements for this target?"… but now I realize that if that were really the case that would have been true for the Intents target as well. I should also have checked with pprof to realize that the Widget's .mobileprovision indeed has the entitlement – which confirms it should be added to the .entitlement file too.

I'm not sure why I said back then that the `.mobileprovision` had the entitlement, because if I check it again I can't find it in the profile anymore, confirming what Gio said that it's not here:

```
$ pprof -i -c 89859073-b915-481d-a6f7-06ff1a0ebad7
- name: match AppStore com.codality.NotationalFlow.Widgets
- uuid: 89859073-b915-481d-a6f7-06ff1a0ebad7
- app_id_name: Simplenote Widgets
- app_id_prefix: ["PZYM8XX95Q"]
- creation_date: 2021-06-02T20:59:09+00:00
- expiration_date: 2022-02-23T04:32:18+00:00
- ttl: 265
- team_ids: ["PZYM8XX95Q"]
- team_name: Automattic, Inc.
- Entitlements:
   - beta-reports-active: true
   - com.apple.security.application-groups: ["group.com.codality.NotationalFlow"]
   - application-identifier: PZYM8XX95Q.com.codality.NotationalFlow.Widgets
   - keychain-access-groups: ["PZYM8XX95Q.*", "com.apple.token"]
   - get-task-allow: false
   - com.apple.developer.team-identifier: PZYM8XX95Q
- 1 Developer Certificates
   - /UID=PZYM8XX95Q/CN=Apple Distribution: Automattic, Inc. (PZYM8XX95Q)/OU=PZYM8XX95Q/O=Automattic, Inc./C=US
- 0 Provisioned Devices
- Provision all devices: false
```

So maybe I confused the files and got lost while juggling with the many different profiles, targets and configs back then, but at least now this + Gio's changes confirm that we don't need it in that target.

We might want to remove it from the `.entitlements` file for the Widget's Debug config too – to align with the other configs of that target, as said in comment below.

@charliescheer Once this last thing is cleaned up and addressed, I say it's ready to :shipit:

**Comment by charliescheer** [comment]
Awesome!  Thanks for all of the help @mokagio  and @AliSoftware I am gonna merge this one now!

**Pull request merged**
